Fix mrack import requiring mrack import

martinhoyer opened the pull request:
Resolves #3716

martinhoyer commented on tmt/steps/provision/mrack.py:
```diff
@@ -1244,24 +1244,37 @@ def is_ready(self) -> bool:
 
-        except mrack.errors.MrackError:
+        except mrack.errors.MrackError as e:
+            self._logger.warning(f"Guest {self.name} job {self.job_id} inspection failed: {e}")
```
does it make sense to do warning?

martinhoyer replied:
Removed

martinhoyer commented:
@coderabbitai full review

happz commented:
> > #3659 is the same issue, just different place.
> 
> So this is something which still needs to be covered?

Rather examined, whether the fix proposed in this PR applies to #3659 as well. I think it does, but since @martinhoyer was knee-deep in fixing this mrack issue, he could be in a better position to find out.

martinhoyer commented:
> > #3659 is the same issue, just different place.
> 
> So this is something which still needs to be covered?

Right, I knew I forgot something. 
@happz @psss I've tried to handle it holistically and it eventually spiraled out of control, where I created _compat/mrack.py, added mrack to banned api and it was just so many changed lines with type checkers going insane. Reverted back to this https://github.com/teemtee/tmt/pull/3718/commits/f2913fa7528825f5a4c681478a884b28ab60fd02, but I'm not sure if it make sense to go this way.

happz commented on tmt/steps/provision/mrack.py:
```diff
@@ -36,14 +36,213 @@
+        """Get mrack errors module"""
+        if not self._imported or self._mrack is None:
+            raise RuntimeError("MrackProxy must be initialized before accessing errors")
+        return self._mrack.errors
```
Both `context` and `errors` could decorated with `@property`, to more resemble the module names in the code below (`mrack_proxy.errors.NotAuthenticatedError`).

martinhoyer replied:
This part is also removed

happz commented:
> > > #3659 is the same issue, just different place.
> > 
> > 
> > So this is something which still needs to be covered?
> 
> Right, I knew I forgot something. @happz @psss I've tried to handle it holistically and it eventually spiraled out of control, where I created _compat/mrack.py, added mrack to banned api and it was just so many changed lines with type checkers going insane. Reverted back to this [f2913fa](https://github.com/teemtee/tmt/commit/f2913fa7528825f5a4c681478a884b28ab60fd02), but I'm not sure if it make sense to go this way.
> 
> edit: I guess it's one of those things where it's better to set it on fire and start again, but don't mind me, I'm just tired :)

Yes, I believe there is another way. I tried with some lazy imports, and I failed. I think there is a chance for some smart solution, but I think we somehow lost the need for a big, generic approach: there used to be more optional imports, today we're down to mrack and testcloud. Which means, it's probably not necessary to spend more time here and now on the grand solution I'd like to see, and improving the situation is good enough. Which I think your patch does, so, LGTM.

psss commented on tmt/steps/provision/mrack.py:
```diff
@@ -36,14 +36,213 @@
+                req['whiteboard'] = host.whiteboard
+
+                # Get logger from proxy
+                proxy_logger = mrack_proxy._logger
```
Similarly here.

psss commented on tmt/steps/provision/mrack.py:
```diff
@@ -36,14 +36,213 @@
+        }
+
+        if name in attr_map:
+            return attr_map[name]
```
What about using an exception instead?

```python
try:
    return attr_map[name]
except KeyError:
    ...
```

martinhoyer replied:
removed

martinhoyer commented on tmt/steps/provision/mrack.py:
```diff
@@ -57,7 +57,26 @@
+    """
+
+    @wraps(func)
+    def wrapper(*args: Any, **kwargs: Any) -> Any:
```
Need to learn about all this, thanks!

psss commented on tmt/steps/provision/mrack.py:
```diff
@@ -57,7 +58,29 @@
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+        if not _MRACK_IMPORTED and _MRACK_IMPORT_ARGS is not None:
+            workdir, name, logger = _MRACK_IMPORT_ARGS
+            import_and_load_mrack_deps(workdir, name, logger)
```
So `_MRACK_IMPORT_ARGS` is defined under `import_and_load_mrack_deps`. This means that this line will be called only if `import_and_load_mrack_deps` was called before. Does that make sense?

martinhoyer replied:
yeah, looks like it's wrong. Giving up.

psss replied:
Ok, then let's use #3754 for now and put this pull request into backlog. We can revisit the approach later.

martinhoyer commented:
closing in favor of https://github.com/teemtee/tmt/pull/3754
